### ISSUE TITLE
Mediapipe with custom tensorflow model predict 

### ISSUE THREAD

**Issue by hailsong** [google/mediapipe#1749]
Hi, I try developing real-time program with mediapipe and my custom tensorflow model. (The model is based on mediapipe's hand recognition solution and it is pre-trained)

I trained my model successful and applied it to mediapipe example.
(it is based on https://google.github.io/mediapipe/solutions/hands.html, hand recognition solution of mediapipe)

But I got a problem with this. I can get high FPS with basic example(around ~30) but, with my model, FPS is fall to around ~13

I guess tensorflow's model.predict() function may not use too much resource... are there issue using tensorflow's model.predict function with mediapipe?

Additionally, I'm using this code below to use my model with mediapipe....

physical_devices = tf.config.list_physical_devices('GPU')
print(physical_devices)
tf.config.experimental.set_memory_growth(physical_devices[0], True)

**Comment by sgowroji** [comment]
@hailsong  Could you share your tensorflow model and platform details to understand the issue much better. Thanks !

**Comment by hailsong** [comment]
@sgowroji
My custom model is simple fully connected layer learning model and its output is softmax to make sure classification. Detail codes is here...

> import pandas as pd
> import numpy as np
> import tensorflow as tf
> from tensorflow import keras
> import matplotlib.pyplot as plt
> 
> print(tf.__version__)
> 
> df = pd.read_csv("Threshold Analysis table.csv")
> print(df.tail())
> size = df.shape[0]
> print(size, "num of data")
> 
> df = df.sample(frac=1).reset_index(drop=True) #shuffle
> 
> test_ratio = 0.2
> test_num = int(size - size * test_ratio)
> 
> train = df[:test_num]
> test = df[test_num:]
> 
> train_x = train[['FA1', 'FA2', 'FA3', 'FA4', 'FA5', 'FD1', 'FD2', 'FD3', 'FD4', 'FD5', 'HA1', 'HA2', 'HA3', 'HA4', 'HA5']].to_numpy()
> train_y = train['real'].to_numpy()

> 
> test_x = test[['FA1', 'FA2', 'FA3', 'FA4', 'FA5', 'FD1', 'FD2', 'FD3', 'FD4', 'FD5', 'HA1', 'HA2', 'HA3', 'HA4', 'HA5']].to_numpy()
> test_y = test['real'].to_numpy()
> 
> print(len(train_y), len(test_y)) #1에서 14사이 정수 label
> 
> model = keras.Sequential([
> keras.layers.Dense(15, activation = 'relu'),
> keras.layers.Dense(25, activation = 'relu'),
> #keras.layers.Dense(30, activation = 'relu'),
> keras.layers.Dense(15, activation='softmax')
> ])
> 
> model.compile(optimizer='adam',
> loss='sparse_categorical_crossentropy',
> metrics=['accuracy'])
> 
> hist = model.fit(train_x, train_y, epochs=100)
> 
> test_loss, test_acc = model.evaluate(test_x,  test_y, verbose=2)
> 
> print('\n테스트 정확도:', test_acc)
> 
> model.save('model_save/my_model.h5')
> print('new model saved')

Dataframe's col names like 'FA1, FA2' .... are relative parameters I get from mediapipe solution.

With this model, I wanna develop real-time classification program. input is 15 numbers that is from mediapipe hand solution(it is relative value group of 3d static hand position). output is estimative class.


model_save/my_model.h5 is opened in another python program, and perform 'predict' function inside while.

I can get real-time classification but it has low FPS and how can I manage FPS...? Thank you 😥

**Comment by hailsong** [comment]
I resolve the issue. Nevertheless I couldn't understand how can tensorflow and mediapipe manage CPU/GPU resource, through multiprocessing and tensorflow.config.experimental.set_memory_growth setting, I solve this.

Firstly, I put set_memory_growth code to start of main function.
(https://www.tensorflow.org/api_docs/python/tf/config/experimental/set_memory_growth)
```
import tensorflow as tf
.
.
.
physical_devices = tf.config.list_physical_devices('GPU')
#print(physical_devices)
tf.config.experimental.set_memory_growth(physical_devices[0], True)
```

And I used multiprocessing module...
```
    shared_array_l = Array('d', [0. for _ in range(15)])
    static_num_l = Value('i', 0)
    shared_array_r = Array('d', [0. for _ in range(15)])
    static_num_r = Value('i', 0)

    process1 = Process(target=main, args=(shared_array_l, static_num_l, shared_array_r, static_num_r))
    process2 = Process(target=process_static_gesture, args=(shared_array_l, static_num_l))
    process3 = Process(target=process_static_gesture, args=(shared_array_r, static_num_r))

    process1.start()
    process2.start()
    process3.start()
    process1.join()
    process2.join()
    process3.join()
```

I guess... mediapipe use tensorflow-lite and using tensorflow-lite with tensorflow model can cause speed reduction...?
I'm going to close this issue... but if there is someone can give me insight, please comment.

Thank you

**Comment by lucky2719** [comment]
> Hi, I try developing real-time program with mediapipe and my custom tensorflow model. (The model is based on mediapipe's hand recognition solution and it is pre-trained)
> 
> I trained my model successful and applied it to mediapipe example.
> (it is based on https://google.github.io/mediapipe/solutions/hands.html, hand recognition solution of mediapipe)
> 
> But I got a problem with this. I can get high FPS with basic example(around ~30) but, with my model, FPS is fall to around ~13
> 
> I guess tensorflow's model.predict() function may not use too much resource... are there issue using tensorflow's model.predict function with mediapipe?
> 
> Additionally, I'm using this code below to use my model with mediapipe....
> 
> physical_devices = tf.config.list_physical_devices('GPU')
> print(physical_devices)
> tf.config.experimental.set_memory_growth(physical_devices[0], True)

Hi can I know how you clubbed tensorflow model with hand mediapipe model is it like training on tensorflow normal model and then giving h5 file as an Input to hand mediapipe model or how is it done thanks in advance
